### PR TITLE
[AIRFLOW-1294] Backfills can loose tasks to execute

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1294

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

In the backfills we can loose tasks to execute due to a task
setting its own state to NONE if concurrency limits are reached,
this makes them fall outside of the scope the backfill is
managing hence they will not be executed.

Several bugs are the cause for this. Firstly, the state
reported by the executor was always reported as success, ie.
the return code of the task instance was not propagated.
Next to that, if the executor already has a task instance
in its queue it will silently ignore the task instance
being added. The backfills did not guard against this, thus
tasks could get lost here as well.

This patch introduces CONCURRENCY_REACHED as an executor
state, which will be set if the task exits with EBUSY (16).
This allows the backfill to properly handle these tasks
and reschedule them. Please note that the CeleryExecutor
does not report back on executor states.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Looking at an approach to updated tests.

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

@aoen @saguziel @mistercrunch

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -1343,6 +1344,7 @@ def run(
             logging.info(msg)
             session.merge(self)
             session.commit()
+            sys.exit(errno.EBUSY)
             return
```
Comment (aoen) [inline airflow/models.py:1348]: return is redundant if we are sys.exiting

File: airflow/executors/local_executor.py [hunk 2]
```diff
@@ -48,9 +49,13 @@ def run(self):
                 subprocess.check_call(command, shell=True)
                 state = State.SUCCESS
             except subprocess.CalledProcessError as e:
-                state = State.FAILED
-                self.logger.error("failed to execute task {}:".format(str(e)))
-                # raise e
+                if e.returncode == errno.EBUSY:
```
Comment (saguziel) [inline airflow/executors/local_executor.py:52]: I really don't think we should have checks in the executor. The executor should only be concerned with executing.

**Comment by aoen** [comment]
I think it should be the responsibility the executor to reschedule tasks to run, with no "special" message passing between task instance run function and executors like sys.exist(EBUSY) (the task should just record it's state in the DB and the executor should use that to determine what to do with the task). Couldn't the job see that the state was None after running the task and reschedule appropriately?

I'm seeing the flow as something like this 
Job -> Executor -> TI.run()
The executor will facilitate the running of the airflow command to run to completion without knowing anything about the task itself or communicating with the Airflow DB. It should be quite dumb. It will be the responsibility of the job to e.g. check TI state is none and re-send it to an executor.

Am I completely off here?

File: airflow/executors/local_executor.py (hunk 2 above)
Comment (bolkedebruin) [reply airflow/executors/local_executor.py:52]: The check is already there, e.g. a non 0 return code make the executor set itself to FAILED.

**Comment by bolkedebruin** [comment]
The current logic is Job -> Executor -> LocalTaskJob -> TI.Run. 

We have Executor and LocalTaskJob to monitor the execution of tasks. You need this kind of logic to guard against tasks that are not able to set themselves to a certain state, e.g. out of memory, sigkill, segfault etc (which are set by the os). Thus here you cannot count on what the state of the TaskInstance is.

As we are executing the tasks as subprocesses and cannot just rely on what the task reports as its state, we will need to rely on how the OS passes messages (errors in this case) between the processes it runs. 

In backfills we are relying on the monitoring (already in 1.7.x) to verify that what the task is reporting as state in congruent with what the executor is thinking the task can report at all. The need for CONCURRENCY_REACHED/EBUSY is to make sure the executor and the task stay in sync. Otherwise the executor would report either success or failed which would not make sense in case of State.NONE (or any other state that does not signal SUCCESS or FAILED). I didn't add CONCURRENCY_REACHED to the task states as that would requiring the Scheduler code to handle that as well. I tried to keep the changes to a minimum.

Obviously I am open to a more simple approach, given the constraints of monitoring. I tried a couple of approaches (removing the TI from the tasks the executor knows when reporting about which was a hack and didn't work), but I ended up with tasks being marked failed incorrectly every single time.

@aoen on your remark that the executor should reschedule tasks I don't think that is smart: it would need its own scheduling logic (backoffs, not scheduling etc). Seems to me as that would severely increase complexity. The executor is quite dumb right now, it just reports back (it already was!) what the OS reports.

**Comment by bolkedebruin** [comment]
Forgot to mention that when I was not using this new executor state I ended up in all kinds Race conditions: the state of the task could already have changed to queued before the executor was able to report back. 

This triggers me to attempt one more simplification

**Comment by bolkedebruin** [comment]
Closing this in favour of the simplified version

**Pull request closed without merging**